cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,474 @@
+        b_advantages = advantages.reshape(-1)
+        b_returns = returns.reshape(-1)
+        b_values = values.reshape(-1)
+
+        # Policy network optimization
+        b_inds = np.arange(args.epoch_size)
+        clipfracs = []
+        for epoch in range(args.policy_update_epochs):
+            np.random.shuffle(b_inds)
+            for start in range(0, args.epoch_size, args.policy_batch_size):
+                end = start + args.policy_batch_size
+                mb_inds = b_inds[start:end]
+
+                _, newlogprob, entropy = agent_policy.get_action(b_obs[mb_inds], b_actions.long()[mb_inds])
+                logratio = newlogprob - b_logprobs[mb_inds]
+                ratio = logratio.exp()
+
+                with torch.no_grad():
+                    # calculate approx_kl http://joschu.net/blog/kl-approx.html
+                    old_approx_kl = (-logratio).mean()
+                    approx_kl = ((ratio - 1) - logratio).mean()
+                    clipfracs += [((ratio - 1.0).abs() > args.clip_coef).float().mean().item()]
+
+                mb_advantages = b_advantages[mb_inds]
+                if args.norm_adv:
+                    mb_advantages = (mb_advantages - mb_advantages.mean()) / (mb_advantages.std() + 1e-8)
+
+                # Policy loss
+                pg_loss1 = -mb_advantages * ratio
+                pg_loss2 = -mb_advantages * torch.clamp(ratio, 1 - args.clip_coef, 1 + args.clip_coef)
+                pg_loss = torch.max(pg_loss1, pg_loss2).mean()
+
+                entropy_loss = entropy.mean()
+                policy_loss = pg_loss - args.ent_coef * entropy_loss
+
+                policy_optimizer.zero_grad()
+                policy_loss.backward()
+                nn.utils.clip_grad_norm_(agent_policy.parameters(), args.max_grad_norm)
+                policy_optimizer.step()
+
+            if args.target_kl is not None:
+                if approx_kl > args.target_kl:
+                    break
+
+        # Value network optimization
+        for epoch in range(args.value_update_epochs):
+            np.random.shuffle(b_inds)
+            for start in range(0, args.epoch_size, args.value_batch_size):
+                end = start + args.value_batch_size
+                mb_inds = b_inds[start:end]
+
+                newvalue = agent_value.get_value(b_obs[mb_inds])
+
+                # Value loss
+                v_loss = 0.5 * ((newvalue - b_returns[mb_inds]) ** 2).mean()
+
+                value_optimizer.zero_grad()
+                v_loss.backward()
+                nn.utils.clip_grad_norm_(agent_value.parameters(), args.max_grad_norm)
+                value_optimizer.step()
+
+        # Value network to policy network distillation
+        agent_policy.zero_grad(True)  # don't clone gradients
+        old_agent_policy = deepcopy(agent_policy)
+        old_agent_policy.eval()
+        for epoch in range(args.distill_update_epochs):
```

Comment:
Just a note, in future work we found by switching the order from Rollout-Value-Policy-Distil, to Rollout-Distil-Policy-Value we could avoid an additional forward (and the copy you need here too) and instead simply use the policy that was generated during rollout. This means the policy updates are slightly off-policy, even so, imperically it works slightly better than the order used in our paper.

Reply:
sounds great, I'll flip the order!